### PR TITLE
fix two bugs in floquet steadystate

### LINKED ISSUE #2386: ENR basis not working with steady state solver (normal and floquet based) in 5.0.0
### Bug Description

Using a Hamiltonian that is constructed using operators returned from the `enr_destroy` function, does not work with the steady state solver in Quito 5.0.0. It does however work with Qutip 4.7 (have not checked any other versions). 

The error given is a dimension mismatch essentially, which seems to stem from an "expected dimension" of the Hamiltonian based on the number of modes. In the example below, the dimension mismatch is between 65536 (which is 2^16, the entire Hilbert space of my problem) vs 17, which is the excitation number restricted subspace. 

### Code to Reproduce the Bug

```shell
N = 16 ## number of resonators including tapers
max_excitation = 1 ## maximum excitation
num_levels = 2
dims = [num_levels] * N

ann_operators = enr_destroy(dims, excitations = max_excitation)
w_r = 2 * np.pi * 5
w = np.array([w_r] * N)

J_r = 2 * np.pi * 0.4
J_t2 = 1.353 * J_r
J_t1 = 1.035 * J_r

J = np.array([J_t2, J_t1] + [J_r] * (N-5) + [J_t1, J_t2])

kappa = 2.92 * J_r
drive_amp = 2 * np.pi * 0.0001
t = []

detuning_list = 2 * np.pi * np.linspace(-1, 1, 100)

for detuning in detuning_list:
    drive_w = (w_r - detuning)

    H_r = 0
    for i, a in enumerate(ann_operators):
        H_r += (w[i] - drive_w) * a.dag() * a

    H_int = 0
    for i in range(len(ann_operators)-1):
        H_int += J[i] * (ann_operators[i].dag() * ann_operators[i+1] + ann_operators[i+1].dag() * ann_operators[i])

    H_d = -1j * drive_amp * (ann_operators[0] - ann_operators[0].dag())

    H = H_r + H_int + H_d

    c_ops = [np.sqrt(kappa)*ann_operators[0], np.sqrt(kappa)*ann_operators[-1]]

    rho_ss = steadystate(H, c_ops)

    a_out = np.sqrt(kappa) * expect(ann_operators[-1], rho_ss)
    t.append(np.sqrt(kappa) * a_out/drive_amp)

print("J = ", np.array(J)/(2 * np.pi))
print("w = ", np.array(w)/(2 * np.pi))
print("kappa = ", kappa/2/np.pi)
t = np.array(t)
plt.plot(detuning_list/2/np.pi, 20 * np.log10(np.abs(t)))
plt.xlabel("Detuning (GHz)")
plt.ylabel("|S21|(dB)")
plt.ylim(-10,0.01)
# plt.xlim(-2, 2)
plt.grid()
```

### Code Output

```shell
{
	"name": "ValueError",
	"message": "Provided dimensions do not match the data: (65536, 65536) vs (17, 17)",
	"stack": "---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[2], line 39
     35 H = H_r + H_int + H_d
     37 c_ops = [np.sqrt(kappa)*ann_operators[0], np.sqrt(kappa)*ann_operators[-1]]
---> 39 rho_ss = steadystate(H, c_ops)
     41 a_out = np.sqrt(kappa) * expect(ann_operators[-1], rho_ss)
     42 t.append(np.sqrt(kappa) * a_out/drive_amp)

File /opt/anaconda3/envs/trial_env/lib/python3.11/site-packages/qutip/solver/steadystate.py:177, in steadystate(A, c_ops, method, solver, **kwargs)
    175     kwargs.pop(\"power_eps\", 0)
    176     kwargs.pop(\"sparse\", 0)
--> 177     return _steadystate_direct(A, kwargs.pop(\"weight\", 0),
    178                                method=solver, **kwargs)
    180 elif method == \"power\":
    181     # Remove unused kwargs, so only used and pass-through ones are included
    182     kwargs.pop(\"weight\", 0)

File /opt/anaconda3/envs/trial_env/lib/python3.11/site-packages/qutip/solver/steadystate.py:246, in _steadystate_direct(A, weight, **kw)
    243 rho_ss = _data.column_unstack(steadystate, n)
    244 rho_ss = _data.add(rho_ss, rho_ss.adjoint()) * 0.5
--> 246 return Qobj(rho_ss, dims=A.dims[0], isherm=True)

File /opt/anaconda3/envs/trial_env/lib/python3.11/site-packages/qutip/core/qobj.py:307, in Qobj.__init__(self, arg, dims, copy, superrep, isherm, isunitary)
    305 self._isherm = isherm
    306 self._isunitary = isunitary
--> 307 self._initialize_data(arg, dims, copy)
    309 if superrep is not None:
    310     self.superrep = superrep

File /opt/anaconda3/envs/trial_env/lib/python3.11/site-packages/qutip/core/qobj.py:300, in Qobj._initialize_data(self, arg, dims, copy)
    296     self._dims = Dimensions(
    297         dims or [[self._data.shape[0]], [self._data.shape[1]]]
    298     )
    299 if self._dims.shape != self._data.shape:
--> 300     raise ValueError('Provided dimensions do not match the data: ' +
    301                      f\"{self._dims.shape} vs {self._data.shape}\")

ValueError: Provided dimensions do not match the data: (65536, 65536) vs (17, 17)"
}
```

### Expected Behaviour

In Qutip 4.7 (and python 3.9), I see the following (correct output):
image

### Your Environment

```shell
QuTiP Version:      5.0.0
Numpy Version:      1.26.4
Scipy Version:      1.12.0
Cython Version:     None
Matplotlib Version: 3.8.4
Python Version:     3.11.8
Number of CPUs:     8
BLAS Info:          Generic
INTEL MKL Ext:      False
Platform Info:      Darwin (x86_64)
```

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by nwlambert**
Two small bugs appeared in floquet steadystate in v5, this is a quick patch.

**Description**
Two bugs:

1) Small minus sign error in one of the terms (comparing to v4.7 and what is written in the old toolbox guide).

2) Some kind of data copying confusion with the S and T operators when updating their data contents with the results.  Just initiating these two separately seems to fix it.

other minor fix: updated the link to the toolbox guide

Some other minor niggles that I didn't fix:
- some of the optional solver methods dont work (seem not to support square right hand sides in '_data.solve()'.  I didn't dig much on fixing these though

- the 'splu' method fails for a couple of reasons. I can get it to work by removing the choice of warnings as errors in 'solve.py' (which make it fail in other instances as well i suppose, like with normal steadystate()), making sure the solver option is not passed to steadystate(), and making sure that S and T are always converted back to CSR() after being updated in the iterative bit (in `steadystate_floquet()' they get returned as dense from _data.solve).  I did not include those changes here though as the docstrings dont actually indicate support for splu anyway, so perhaps not needed.

I am marking this as draft at the moment as I also want to update the test to be sensitive to these bugs (its a bit too coarse at the moment), and double check that the results from this solver are now correct.

**Related issues or PRs**
fix #2386

### CODE REVIEW

**Comment by nwlambert** [comment]
I adjusted the test to be sensitive to the original issues.  new test is just a driven two-level system. not sure why the old test of a driven boson was insensitive to that test. i could make it sensitive by making the drive act on the frequency, like the example in #2386 but this is simpler. 

also adjusted the test to compare a period of the dynamics from mesolve to the result from the steadystate solver.

File: qutip/solver/steadystate.py [hunk 1]
```diff
@@ -373,18 +374,20 @@ def steadystate_floquet(H_0, c_ops, Op_t, w_d=1.0, n_it=3, sparse=False,
...
-    Section (10.16)
+    https://painterlab.caltech.edu/wp-content/uploads/2019/06/qe_quantum_optics_toolbox.pdf,
+    Section (16)
 
     """
 
     L_0 = liouvillian(H_0, c_ops)
-    L_m = L_p = 0.5 * liouvillian(Op_t)
+    L_m = 0.5 * liouvillian(Op_t)
+    L_p = 0.5 * liouvillian(Op_t)
```
Comment (hodgestar) [inline qutip/solver/steadystate.py:384]: I'm a bit surprised this makes a difference. Are `L_m` or `L_p` modified somewhere?

**Review by hodgestar — COMMENTED** [review]
Changes look good so far!

File: qutip/solver/steadystate.py (hunk 1 above)
Comment (nwlambert) [reply qutip/solver/steadystate.py:384]: ah no, it was just S and T, but did the same here out of abundance of caution when i was trying to find out what was going wrong

**Review by Ericgig — APPROVED** [review]

**Pull request merged**